compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -387,24 +386,31 @@ pub(crate) fn target_features_cfg(sess: &Session, allow_unstable: bool) -> Vec<S
         }
     }
 
-    // Filter enabled features based on feature gates
-    sess.target
-        .rust_target_features()
-        .iter()
-        .filter_map(|(feature, gate, _)| {
-            // The `allow_unstable` set is used by rustc internally to determined which target
-            // features are truly available, so we want to return even perma-unstable "forbidden"
-            // features.
-            if allow_unstable
-                || (gate.in_cfg() && (sess.is_nightly_build() || gate.requires_nightly().is_none()))
-            {
-                Some(Symbol::intern(feature))
-            } else {
-                None
-            }
-        })
-        .filter(|feature| features.contains(&feature))
-        .collect()
+    // Filter enabled features based on feature gates.
+    let f = |allow_unstable| {
+        sess.target
+            .rust_target_features()
+            .iter()
+            .filter_map(|(feature, gate, _)| {
+                // The `allow_unstable` set is used by rustc internally to determined which target
+                // features are truly available, so we want to return even perma-unstable
+                // "forbidden" features.
+                if allow_unstable
+                    || (gate.in_cfg()
+                        && (sess.is_nightly_build() || gate.requires_nightly().is_none()))
+                {
+                    Some(Symbol::intern(feature))
+                } else {
+                    None
+                }
+            })
+            .filter(|feature| features.contains(&feature))
+            .collect()
+    };
```

Comment:
The code in this closure is duplicated between all codegen backends. Maybe move it to the caller of `target_features_cfg`? So just return `features` from `target_features_cfg` and then do the splitting between stable and unstable features in the caller.

Reply:
I don't see it in the cranelift backend.

Reply:
And if the response is "the cranelift backend *should* do that filtering", that would also be a good follow-up, because that's a pre-existing issue that is distinct from the one this PR is addressing.

Reply:
The Cranelift backend currently doesn't consider any unstable feature to ever be enabled at all, so filtering has no effect. Once it does support unstable features, yes it should do filtering.

Reply:
Ok, makes sense.